subprojects/core/src/main/java/org/gradle/initialization/ParallelismBuildOptionFactory.java

```diff
@@ -0,0 +1,80 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.initialization;
+
+import org.gradle.concurrent.ParallelismConfiguration;
+import org.gradle.internal.buildoption.BooleanBuildOption;
+import org.gradle.internal.buildoption.BuildOption;
+import org.gradle.internal.buildoption.BuildOptionFactory;
+import org.gradle.internal.buildoption.CommandLineOptionConfiguration;
+import org.gradle.internal.buildoption.StringBuildOption;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class ParallelismBuildOptionFactory implements BuildOptionFactory<ParallelismConfiguration> {
+
+    private final List<BuildOption<ParallelismConfiguration>> options = new ArrayList<BuildOption<ParallelismConfiguration>>();
+
+    public ParallelismBuildOptionFactory() {
+        options.add(new ParallelOption());
+        options.add(new MaxWorkersOption());
+    }
+
+    @Override
+    public List<BuildOption<ParallelismConfiguration>> create() {
+        return options;
+    }
+
+    public static class ParallelOption extends BooleanBuildOption<ParallelismConfiguration> {
+        public static final String GRADLE_PROPERTY = "org.gradle.parallel";
+
+        public ParallelOption() {
+            super(GRADLE_PROPERTY, CommandLineOptionConfiguration.create("parallel", "Build projects in parallel. Gradle will attempt to determine the optimal number of executor threads to use.").incubating());
+        }
+
+        @Override
+        public void applyTo(boolean value, ParallelismConfiguration settings) {
+            settings.setParallelProjectExecutionEnabled(value);
+        }
+    }
+
+    public static class MaxWorkersOption extends StringBuildOption<ParallelismConfiguration> {
+        public static final String GRADLE_PROPERTY = "org.gradle.workers.max";
+
+        public MaxWorkersOption() {
+            super(GRADLE_PROPERTY, CommandLineOptionConfiguration.create("max-workers", "Configure the number of concurrent workers Gradle is allowed to use.").incubating());
+        }
+
+        @Override
+        public void applyTo(String value, ParallelismConfiguration settings) {
+            try {
+                int workerCount = Integer.parseInt(value);
+                if (workerCount < 1) {
+                    invalidMaxWorkersSwitchValue(value);
+                }
+                settings.setMaxWorkerCount(workerCount);
+            } catch (NumberFormatException e) {
+                invalidMaxWorkersSwitchValue(value);
+            }
+        }
+
+        private void invalidMaxWorkersSwitchValue(String value) {
+            throw new IllegalArgumentException(String.format("Argument value '%s' given for system property %s or --%s option is invalid (must be a positive, non-zero, integer)", value, gradleProperty, commandLineOptionConfiguration.getLongOption()));
```

Comment:
This shouldn't really have an 'or' in the message. We know exactly whether the value came from a system property or command-line parameter. The 'or' just makes things confusing for the user.

Reply:
The `apply` method would have to know what the origin of the value is, either the properties or the command line option. We can change the method.